File: app/services/auth/verification_service.py
"""VerificationService — orchestrates verification code lifecycle.

Responsibilities:
  - Generate 6-digit codes (or reset tokens)
  - Persist via VerificationCodeRepository
  - Enforce rate limits (per-target cooldown + per-uid window)
  - Send via EmailService
  - Verify user-supplied codes (single-use, expiry, brute-force cap)
  - Orchestrate multi-step auth flows (change password with 2FA, bind email,
    reset password with token) — owns the transaction boundary for these.

Valid purposes:
  - bind_email     — bind/change email (requires login)
  - reset_password — forgot password flow (public)
  - twofa          — sensitive operation second factor (requires login)

This service does NOT know about HTTP or sessions; it raises ValueError
on business-rule violations and the router converts to HTTPException.

Transaction boundary: standalone write methods (send_code, send_reset_token)
own their transaction via transactional_scope(). Composable methods
(verify_code, consume_code) accept a db and participate in the caller's
transaction. Multi-service orchestration methods own the transaction and
instantiate collaborating services (UserService) with the tx session.
"""
from __future__ import annotations

import secrets
import string
from datetime import datetime, timedelta, timezone
from typing import Any, Optional

from loguru import logger
from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import settings
from app.infra.transaction import transactional_scope
from app.models import VerificationCode
from app.repository.verification_code_repository import (
    get_verification_code_repository,
)
from app.repository.user_repository import get_user_repository
from app.services.auth.user_service import UserService
from app.services.email_service import send_verification_code, EmailServiceError


VALID_PURPOSES = {"bind_email", "reset_password", "twofa"}


class VerificationService:
    """Stateless orchestration of email verification codes.

    Methods accept db per-call; no instance session state. Standalone
    writes own their transaction; composable methods participate in the
    caller's transaction.
    """

    def __init__(self) -> None:
        self.repo = get_verification_code_repository()

    # ── Send (standalone — owns tx) ────────────────────────────────

    async def send_code(
        self,
        db: AsyncSession,
        *,
        uid: int,
        target: str,
        purpose: str,
    ) -> None:
        """Generate, persist, and email a verification code.

        Rate limits:
          - Same target: at most 1 send per `rate_limit_target_seconds` (default 60s)
          - Same uid: at most `rate_limit_uid_max` sends in
            `rate_limit_uid_minutes` window (default 5 / 10 min)

        Raises ValueError on rate-limit or email failure.
        """
        if purpose not in VALID_PURPOSES:
            raise ValueError(f"无效的验证用途: {purpose}")
        if not target:
            raise ValueError("邮箱不能为空")

        code = _generate_numeric_code(settings.email_code_length)
        async with transactional_scope() as tx_db:
            await self._enforce_rate_limits(tx_db, uid, target)
            # Void any prior unused codes for same target+purpose so only the
            # latest one is valid.
            await self.repo.void_recent_unused(
                tx_db, target=target, purpose=purpose
            )
            await self.repo.create(
                tx_db,
                uid=uid,
                target=target,
                purpose=purpose,
                code=code,
                ttl_seconds=settings.email_code_ttl_seconds,
            )

        # Email send is outside tx — can't roll back, and a sent email with
        # a persisted code is recoverable (user can retry, rate limit gates).
        try:
            await send_verification_code(target, code, purpose)
        except EmailServiceError as e:
            logger.warning(
                "[VERIFY] email send failed uid=%s target=%s purpose=%s err=%s",
                uid, target, purpose, e,
            )
            raise ValueError(str(e)) from e

        logger.info(
            "[VERIFY] code sent uid=%s target=%s purpose=%s",
            uid, target, purpose,
        )

    async def send_reset_token(self, db: AsyncSession, *, target: str) -> None:
        """Public entry: email a reset token to unverified user.

        Looks up uid by email; raises ValueError if no user has that email.
        Uses a 32-byte URL-safe token instead of a 6-digit code so it can
        be embedded in a reset link.
        """
        user = await get_user_repository().find_by_email(target, db)
        if not user:
            # Don't leak whether the email is registered.
            raise ValueError("如果该邮箱已注册，您将收到重置邮件")

        token = secrets.token_urlsafe(32)
        async with transactional_scope() as tx_db:
            await self._enforce_rate_limits(tx_db, user.uid, target)
            await self.repo.void_recent_unused(
                tx_db, target=target, purpose="reset_password"
            )
            # Reset tokens live 10 minutes (longer than 6-digit codes).
            await self.repo.create(
                tx_db,
                uid=user.uid,
                target=target,
                purpose="reset_password",
                code=token,
                ttl_seconds=600,
            )

        try:
            await send_verification_code(target, token, "reset_password")
        except EmailServiceError as e:
            raise ValueError(str(e)) from e

        logger.info(
            "[VERIFY] reset token sent uid=%s target=%s",
            user.uid, target,
        )

    # ── Verify (composable — participates in caller tx) ────────────

    async def verify_code(
        self,
        db: AsyncSession,
        *,
        uid: int,
        target: str,
        purpose: str,
        code: str,
    ) -> int:
        """Verify a user-supplied code WITHOUT consuming it.

        Returns the verification_code.id on success. Caller MUST call
        ``consume_code(db, vc_id)`` after the downstream business operation
        (e.g. password change, email bind) succeeds, so that a failure in
        the business step leaves the code reusable for retry.

        Raises ValueError on:
          - no matching code found (expired or never sent)
          - wrong code (also increments attempt counter)
          - brute-force cap reached (code auto-voided)
        """
        vc = await self.repo.find_latest_unused(
            db, target=target, purpose=purpose
        )
        if vc is None:
            raise ValueError("验证码已过期或未发送，请重新获取")
        if vc.uid != uid:
            # Code was issued to a different uid — treat as invalid.
            raise ValueError("验证码无效")

        if vc.attempts and vc.attempts >= settings.email_max_verify_attempts:
            # Cap reached — void the code so it can't be retried.
            await self.repo.mark_used(db, vc.id)
            logger.warning(
                "[VERIFY] brute-force cap reached target={} purpose={}",
                target, purpose,
            )
            raise ValueError("验证码错误次数过多，请重新获取")

        if vc.code != code:
            attempts = await self.repo.increment_attempts(db, vc.id)
            logger.info(
                "[VERIFY] wrong code target={} purpose={} attempts={}",
                target, purpose, attempts,
            )
            raise ValueError("验证码不正确")

        logger.info(
            "[VERIFY] ok (not yet consumed) uid={} target={} purpose={} vc_id={}",
            uid, target, purpose, vc.id,
        )
        return vc.id

    async def consume_code(self, db: AsyncSession, vc_id: int) -> None:
        """Mark a verified code as used. Call this only AFTER the downstream
        business operation (password change, email bind, etc.) succeeds."""
        await self.repo.mark_used(db, vc_id)
        logger.info("[VERIFY] consumed vc_id={}", vc_id)

    async def verify_reset_token(
        self, db: AsyncSession, *, token: str
    ) -> int:
        """Verify a reset token, return the uid it was issued to.

        Public entry — caller is responsible for enforcing that the token
        came from an email link (not user-supplied without proof).
        """
        now = datetime.now(timezone.utc)
        result = await db.execute(
            select(VerificationCode)
            .where(
                VerificationCode.purpose == "reset_password",
                VerificationCode.code == token,
                VerificationCode.used.is_(False),
                VerificationCode.expires_at > now,
            )
            .order_by(VerificationCode.created_at.desc())
            .limit(1)
        )
        vc = result.scalar_one_or_none()
        if vc is None:
            raise ValueError("重置链接无效或已过期")
        return vc.uid

    async def consume_reset_token(
        self, db: AsyncSession, *, token: str
    ) -> int:
        """Verify + mark used. Returns uid for the caller to update password."""
        uid = await self.verify_reset_token(db, token=token)
        await db.execute(
            update(VerificationCode)
            .where(
                VerificationCode.code == token,
                VerificationCode.purpose == "reset_password",
            )
            .values(used=True)
        )
        return uid

    # ── Multi-service orchestration (owns tx) ──────────────────────

    async def verify_and_change_password(
        self,
        db: AsyncSession,
        *,
        uid: int,
        old_password: str,
        new_password: str,
        email_code: Optional[str],
        sf: Any,
    ) -> None:
        """Change password with optional 2FA verification.

        If the user has a verified email, requires an email_code (2FA).
        Verifies code → changes password → consumes code, atomically.
        """
        async with transactional_scope() as tx_db:
            user = await get_user_repository().get_by_uid(uid, tx_db)
            vc_id: Optional[int] = None
            if user and user.email_verified and user.email:
                if not email_code:
                    raise ValueError("修改密码需要邮箱验证码")
                vc_id = await self.verify_code(
                    tx_db,
                    uid=uid,
                    target=user.email,
                    purpose="twofa",
                    code=email_code,
                )
            user_service = UserService(tx_db, sf)
            await user_service.change_password(uid, old_password, new_password)
            if vc_id is not None:
                await self.consume_code(tx_db, vc_id)

    async def verify_and_bind_email(
        self,
        db: AsyncSession,
        *,
        uid: int,
        email: str,
        code: str,
        purpose: str,
        sf: Any,
    ) -> None:
        """Verify email code and bind email (purpose=bind_email), or just
        verify (purpose=twofa). Atomically verifies → binds → consumes."""
        async with transactional_scope() as tx_db:
            vc_id = await self.verify_code(
                tx_db, uid=uid, target=email, purpose=purpose, code=code,
            )
            if purpose == "bind_email":
                user_service = UserService(tx_db, sf)
                await user_service.apply_verified_email(uid, email)
            # Consume the code only after the downstream bind succeeds.
            await self.consume_code(tx_db, vc_id)

    async def consume_token_and_reset_password(
        self,
        db: AsyncSession,
        *,
        token: str,
        new_password: str,
        sf: Any,
    ) -> None:
        """Consume reset token and reset password, atomically."""
        async with transactional_scope() as tx_db:
            uid = await self.consume_reset_token(tx_db, token=token)
            user_service = UserService(tx_db, sf)
            await user_service.reset_password(uid, new_password)

    # ── Rate limit enforcement ────────────────────────────────────

    async def _enforce_rate_limits(
        self, db: AsyncSession, uid: int, target: str
    ) -> None:
        """Raise ValueError if rate limits would be exceeded."""
        now = datetime.now(timezone.utc)

        # Per-target cooldown: latest code within N seconds blocks new sends.
        recent_target_count = await self.repo.count_recent_by_target(
            db,
            target=target,
            since=now - timedelta(seconds=settings.email_rate_limit_target_seconds),
        )
        if recent_target_count > 0:
            raise ValueError(
                f"请求过于频繁，请 {settings.email_rate_limit_target_seconds} 秒后重试"
            )

        # Per-uid window: at most N sends in M minutes.
        recent_uid_count = await self.repo.count_recent_by_uid(
            db,
            uid=uid,
            since=now - timedelta(minutes=settings.email_rate_limit_uid_minutes),
        )
        if recent_uid_count >= settings.email_rate_limit_uid_max:
            raise ValueError(
                "请求过于频繁，请稍后重试"
            )


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _generate_numeric_code(length: int) -> str:
    """Cryptographically-random numeric code (avoids 0/O ambiguity not needed
    because it's all digits)."""
    alphabet = string.digits
    return "".join(secrets.choice(alphabet) for _ in range(length))

File: app-go/app-auth/internal/service/verification_service.go
package service

import (
	"context"
	"crypto/rand"
	"errors"
	"fmt"
	"log/slog"
	"math/big"
	"regexp"
	"time"

	"app-auth/internal/config"
	"app-auth/internal/model"
	"app-auth/internal/repo"

	"gorm.io/gorm"
)

// Verification — orchestrates the verification-code lifecycle:
// generate → persist (per-target cooldown + per-uid window) → deliver
// (Resend email / Aliyun SMS) → verify (single-use, brute-force cap) →
// multi-step flows (password change with 2FA, email bind, password reset).
// Port of the Python VerificationService.
type Verification struct {
	cfg   *config.Config
	email *EmailService
	sms   *SMSService
	auth  *AuthService
	repo  *repo.VerificationRepo
	users *repo.UserRepo
}

func NewVerification(cfg *config.Config, email *EmailService, sms *SMSService, auth *AuthService) *Verification {
	return &Verification{
		cfg:   cfg,
		email: email,
		sms:   sms,
		auth:  auth,
		repo:  repo.NewVerificationRepo(),
		users: repo.NewUserRepo(),
	}
}

var phoneRe = regexp.MustCompile(`^1[3-9]\d{9}$`)

var (
	emailPurposes = map[string]bool{"bind_email": true, "reset_password": true, "twofa": true, "register": true}
	smsPurposes   = map[string]bool{"login": true, "bind_phone": true, "twofa_sms": true}
)

var errInvalidPurpose = errors.New("无效的验证用途")

func (v *Verification) channel(purpose string) (string, bool) {
	if emailPurposes[purpose] {
		return "email", true
	}
	if smsPurposes[purpose] {
		return "sms", true
	}
	return "", false
}

// SendCode generates, persists and delivers a code for (uid, target, purpose).
// uid may be 0 for pre-registration flows.
func (v *Verification) SendCode(ctx context.Context, uid int64, target, purpose string) error {
	channel, ok := v.channel(purpose)
	if !ok {
		return errInvalidPurpose
	}
	if target == "" {
		return errors.New("接收方不能为空")
	}
	if channel == "sms" && !phoneRe.MatchString(target) {
		return errors.New("手机号格式不正确")
	}
	db := v.auth.DB()
	if err := v.enforceRateLimits(db, uid, target); err != nil {
		return err
	}
	code, err := numericCode(v.cfg.Email.CodeLength)
	if err != nil {
		return err
	}
	now := time.Now()
	var uidPtr *int64
	if uid != 0 {
		uidPtr = &uid
	}
	if err := db.Transaction(func(tx *gorm.DB) error {
		if err := v.repo.InvalidateOlder(tx, target, purpose); err != nil {
			return err
		}
		return v.repo.Create(tx, &model.VerificationCode{
			UID:       uidPtr,
			Target:    target,
			Type:      channel,
			Purpose:   purpose,
			Code:      code,
			ExpiresAt: now.Add(time.Duration(v.cfg.Email.CodeTTLSeconds) * time.Second),
			Attempts:  intPtr(0),
			CreatedAt: &now,
		})
	}); err != nil {
		return err
	}

	// Delivery outside the tx — a sent message with a persisted code is
	// recoverable; the rate limit gates retries (parity with Python).
	if channel == "sms" {
		if err := v.sms.SendSms(ctx, target, code); err != nil {
			slog.Warn("[VERIFY] sms send failed", "uid", uid, "purpose", purpose, "err", err)
			return err
		}
	} else {
		if err := v.email.SendVerificationCode(ctx, target, code, purpose); err != nil {
			slog.Warn("[VERIFY] email send failed", "uid", uid, "purpose", purpose, "err", err)
			return err
		}
	}
	slog.Info("[VERIFY] code sent", "channel", channel, "uid", uid, "purpose", purpose)
	return nil
}

// SendResetToken emails a 32-byte reset token (10-minute TTL) to a
// registered address, without leaking whether the email exists.
func (v *Verification) SendResetToken(ctx context.Context, target string) error {
	db := v.auth.DB()
	user, err := v.users.GetByEmail(db, target)
	if err != nil {
		return err
	}
	if user == nil {
		// Do not leak account existence; the router returns a success-shaped
		// message regardless.
		return errors.New("如果该邮箱已注册，您将收到重置邮件")
	}
	if err := v.enforceRateLimits(db, user.UID, target); err != nil {
		return err
	}
	token, err := randomToken(32)
	if err != nil {
		return err
	}
	now := time.Now()
	if err := db.Transaction(func(tx *gorm.DB) error {
		if err := v.repo.InvalidateOlder(tx, target, "reset_password"); err != nil {
			return err
		}
		return v.repo.Create(tx, &model.VerificationCode{
			UID:       &user.UID,
			Target:    target,
			Type:      "email",
			Purpose:   "reset_password",
			Code:      token,
			ExpiresAt: now.Add(600 * time.Second),
			Attempts:  intPtr(0),
			CreatedAt: &now,
		})
	}); err != nil {
		return err
	}
	if err := v.email.SendVerificationCode(ctx, target, token, "reset_password"); err != nil {
		return err
	}
	slog.Info("[VERIFY] reset token sent", "uid", user.UID)
	return nil
}

// VerifyCode checks a user-supplied code WITHOUT consuming it; returns the
// verification_codes row id. Caller must ConsumeCode after the business
// step succeeds so a failed business step leaves the code retryable.
func (v *Verification) VerifyCode(target, purpose, code string, uid int64) (int64, error) {
	db := v.auth.DB()
	vc, err := v.repo.FindLatestUnused(db, target, purpose)
	if err != nil {
		return 0, err
	}
	if vc == nil {
		return 0, errors.New("验证码已过期或未发送，请重新获取")
	}
	if uid != 0 && vc.UID != nil && *vc.UID != uid {
		return 0, errors.New("验证码无效")
	}
	attempts := 0
	if vc.Attempts != nil {
		attempts = *vc.Attempts
	}
	if attempts >= v.cfg.Email.MaxVerifyAttempts {
		_ = v.repo.MarkUsed(db, vc.ID)
		slog.Warn("[VERIFY] brute-force cap reached", "target", target, "purpose", purpose)
		return 0, errors.New("验证码错误次数过多，请重新获取")
	}
	if vc.Code != code {
		newAttempts, _ := v.repo.BumpAttempts(db, target, purpose, vc.Code)
		slog.Info("[VERIFY] wrong code", "purpose", purpose, "attempts", newAttempts)
		return 0, errors.New("验证码不正确")
	}
	return vc.ID, nil
}

// ConsumeCode marks a verified code used (after the business step succeeds).
func (v *Verification) ConsumeCode(id int64) error {
	return v.repo.MarkUsed(v.auth.DB(), id)
}

// VerifyAndChangePassword — change password with optional email 2FA
// (required when a verified email exists). Verify → change → consume.
func (v *Verification) VerifyAndChangePassword(uid int64, oldPassword, newPassword, emailCode string) error {
	db := v.auth.DB()
	user, err := v.users.GetByUID(db, uid)
	if err != nil {
		return err
	}
	var vcID int64
	if user != nil && user.EmailVerified != nil && *user.EmailVerified && user.Email != nil {
		if emailCode == "" {
			return errors.New("修改密码需要邮箱验证码")
		}
		vcID, err = v.VerifyCode(*user.Email, "twofa", emailCode, uid)
		if err != nil {
			return err
		}
	}
	if err := v.auth.ChangePassword(uid, oldPassword, newPassword); err != nil {
		return err
	}
	if vcID != 0 {
		if err := v.ConsumeCode(vcID); err != nil {
			return err
		}
	}
	return nil
}

// VerifyAndSetPassword — first-time password set with a mandatory second
// factor: email code when a verified email exists, SMS code when only a
// verified phone exists, otherwise refused (parity with the Python flow —
// a bare session must not be able to plant a persistent credential).
func (v *Verification) VerifyAndSetPassword(uid int64, newPassword, emailCode, smsCode string) error {
	db := v.auth.DB()
	user, err := v.users.GetByUID(db, uid)
	if err != nil {
		return err
	}
	if user == nil {
		return ErrUserNotFound
	}
	var vcID int64
	switch {
	case user.EmailVerified != nil && *user.EmailVerified && user.Email != nil:
		if emailCode == "" {
			return errors.New("设置密码需要邮箱验证码")
		}
		vcID, err = v.VerifyCode(*user.Email, "twofa", emailCode, uid)
		if err != nil {
			return err
		}
	case user.PhoneVerified != nil && *user.PhoneVerified && user.Phone != nil:
		if smsCode == "" {
			return errors.New("设置密码需要短信验证码")
		}
		vcID, err = v.VerifyCode(*user.Phone, "twofa_sms", smsCode, uid)
		if err != nil {
			return err
		}
	default:
		return errors.New("请先绑定并验证邮箱或手机号，再设置密码")
	}
	if err := v.auth.SetPassword(uid, newPassword); err != nil {
		return err
	}
	return v.ConsumeCode(vcID)
}

// VerifyAndBindEmail — verify code and bind (bind_email) or just verify
// (twofa). Order: verify → bind → consume (consume only after bind succeeds).
func (v *Verification) VerifyAndBindEmail(uid int64, email, code, purpose string) error {
	vcID, err := v.VerifyCode(email, purpose, code, uid)
	if err != nil {
		return err
	}
	if purpose == "bind_email" {
		if err := v.auth.ApplyVerifiedEmail(uid, email); err != nil {
			return err
		}
	}
	return v.ConsumeCode(vcID)
}

// VerifyAndBindPhone — verify SMS code and bind the phone (login state).
func (v *Verification) VerifyAndBindPhone(uid int64, phone, code string) error {
	vcID, err := v.VerifyCode(phone, "bind_phone", code, uid)
	if err != nil {
		return err
	}
	if err := v.auth.ApplyVerifiedPhone(uid, phone); err != nil {
		return err
	}
	return v.ConsumeCode(vcID)
}

// ConsumeTokenAndResetPassword — verify + burn the reset token, then reset
// the password. Atomically ordered: burn first (single use), then reset.
func (v *Verification) ConsumeTokenAndResetPassword(token, newPassword string) error {
	db := v.auth.DB()
	vc, err := v.repo.FindLatestUnusedResetToken(db, token)
	if err != nil {
		return err
	}
	if vc == nil || vc.UID == nil {
		return errors.New("重置链接无效或已过期")
	}
	uid := *vc.UID
	if err := v.repo.MarkUsedResetToken(db, token); err != nil {
		return err
	}
	return v.auth.ResetPassword(uid, newPassword)
}

// enforceRateLimits raises when the per-target cooldown or per-uid window
// would be exceeded (data comes from verification_codes, MySQL-backed).
func (v *Verification) enforceRateLimits(db *gorm.DB, uid int64, target string) error {
	now := time.Now()
	n, err := v.repo.CountRecentByTarget(db, target,
		now.Add(-time.Duration(v.cfg.Email.RateLimitTargetSeconds)*time.Second))
	if err != nil {
		return err
	}
	if n > 0 {
		return fmt.Errorf("请求过于频繁，请 %d 秒后重试", v.cfg.Email.RateLimitTargetSeconds)
	}
	if uid == 0 {
		return nil
	}
	m, err := v.repo.CountRecentByUID(db, uid,
		now.Add(-time.Duration(v.cfg.Email.RateLimitUIDMinutes)*time.Minute))
	if err != nil {
		return err
	}
	if m >= int64(v.cfg.Email.RateLimitUIDMax) {
		return errors.New("请求过于频繁，请稍后重试")
	}
	return nil
}

// numericCode generates a cryptographically random numeric code.
func numericCode(length int) (string, error) {
	out := make([]byte, length)
	for i := range out {
		n, err := rand.Int(rand.Reader, big.NewInt(10))
		if err != nil {
			return "", err
		}
		out[i] = byte('0' + n.Int64())
	}
	return string(out), nil
}

func intPtr(i int) *int { return &i }
